### ISSUE TITLE
RFC: Amplify Library Modularization and Bundle Size Improvement

### ISSUE THREAD

**Issue by powerful23** [aws-amplify/amplify-js#3365]
This issue is a Request For Comments (RFC). It is intended to elicit community feedback regarding a proposed change to the library. Please feel free to post comments or questions here.

## Situation

The bundle size of amplify-js has always been a concern to the customers.  At the very beginning we put all the categories in a single package which leads to a huge bundle size. Then we decided to split the whole package into several ones based on the category so we can import each category on need.

## The Problem

Although the bundle size can be largely reduced by using scoped packages instead of directly using `aws-amplify`, there are still many things that we can do to make it better:

- The way of using scoped packages is tricky and not intuitive.
- When using `aws-amplify-react`, all the Amplify library will be imported and bundled regardless of the actual usage.
- Using default export is not the best practice.
- Graphql will be imported when using API module which is unnecessary if only calling REST APIs.
- Version conflicting when using scoped packages which will lead to multiple versions of same package under `node_modules`
- The `aws-sdk` takes a lot of the space in the bundle size but we are only using part of it.

## Proposed Solutions

- Generate ES6 modules so that bundlers like Webpack and Rollup can do the tree shaking when importing something from `aws-amplify`. So instead of using scoped packages like `import Auth from '@aws-amplify/auth'` you can now just use `import { Auth } from 'aws-amplify'` and the bundler will do the rest work for you.
- Update the library code and the documentation to stop using default import/export but also keep backward compatibility.
- Split the API module into two, one will be only for REST APIs, the other one will be for GraphQL APIs. For example, `import { RestAPI, GraphQLAPI } from 'aws-amplify'`
- Use peer dependencies and loose version number to avoid version conflicting.
- Build a http client to replace `aws-sdk`. The client should be able to generate the request context, sign the request with AWS credentials, validate the input parameters, configure the retry policy etc.

## Related Issues

#122 
#2781 
#3097 
#3096 
#2309 
#2199
#1831 
#701 
#2894 
#2058
#488

## Reference

- [Tree shaking](https://webpack.js.org/guides/tree-shaking/)
- [Static module structure](http://exploringjs.com/es6/ch_modules.html#static-module-structure)
- [Why we have banned default exports in Javascript and you should do the same
](https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad)
- [How to Create a React app from scratch using Webpack 4](https://www.freecodecamp.org/news/part-1-react-app-from-scratch-using-webpack-4-562b1d231e75/)

Thanks to @aldo-roman for the working of #2313 

### 05/31/19

You can now use the preview version of amplify to test:
```
npm install aws-amplify@preview
```
The preview version has generated the ESM modules for all the categories so for example, if you are using webpack 4 bundling your react app, and you are only using `Auth` module like:
```js
import { Auth } from 'aws-amplify';

Auth.signIn();
```
Then only the `Auth` module will be included in the bundle. Note that the tree shaking will only happen under the production mode.

**Comment by tuhinkarmakar** [comment]
I tried the preview versions with my Angular app. The reduction in bundle size was significant. However the setup process wasn't easy. I had to go through a lot of trial and error to make it work.

**Steps**

1. Import the modules from the `aws-amplify` package in `main.ts`.
2. Configure the modules individually i.e. `Auth.configure(config)`
3. Register the module with the `Amplify` module using `Amplify.register(Auth)`
4. Add the `AmplifyService` to your root module as `{ provide: AmplifyService, useFactory: () => AmplifyModules({ Auth })`.

**Comment by powerful23** [comment]
@tuhinkarmakar Hi, thanks for testing! I think Step 1, 2, 4 are required but Step 3 is not. Do you have any issue if not calling `Amplify.register()`?

**Comment by tuhinkarmakar** [comment]
@powerful23 Without step 3, I was getting an error saying "No Auth module registered with Amplify". I updated my comment to add 2 more steps that were required.

**Comment by mydea** [comment]
Even when just pulling in `@aws-amplify/auth`, it adds ~80kb of gzipped size. That does feel pretty extensive for just handling authentication. (See https://bundlephobia.com/result?p=@aws-amplify/auth@1.2.24, which also matches what I can see in my app). Are there any plans (or ways) to reduce this? It seems that almost 50% of this is `aws-sdk`. Is there no way to use Cognito without adding on 270kb (un-gzipped) of JS to your app?

**Comment by aldo-roman** [comment]
@mydea it seems `aws-sdk` is publishing only on UMD and, thus, does not support tree-shaking. So i guess it is not possible with current status of the library. They should go through a similar work proposed on this issue. See https://github.com/aws/aws-sdk-js/issues/1766

**Comment by nihakue** [comment]
Does this: https://aws.amazon.com/blogs/developer/new-aws-sdk-for-javascript-developer-preview/ influence the need for writing your own http client? If you can import just the parts of the sdk you need, you can significantly reduce the bundle size.

**Comment by mydea** [comment]
If I try to force using `aws-sdk@latest` it just gives me `2.329` and it remains the same output size!

**Comment by mainawycliffe** [comment]
Impressive, just shaved off around 500K from 1.68 to 1.17, still this library is taking between 30 - 35% of the remaining 1.17 and i only need Authentication. I am using the preview of Angular. 

I see the angular components inside the final bundle and am not using them, i have my own custom components. And they are occupying a whooping 5% on their own. For some tree shaking is not removing them for some reason.

**Comment by jadbox** [comment]
I must echo @ahallora, GraphQL is one of the largest bundled packages in production app... yet we don't use it at all. It's a massive tax when using aws-amplify to the point where we're considering forking @aws-amplify/api just to strip it out.

**Comment by jadbox** [comment]
@powerful23 thought I'd link your PR here for splitting the API and Graphql features:
https://github.com/aws-amplify/amplify-js/pull/3362

**Comment by roborourke** [comment]
This shaved off 180kb from my bundle of just analytics and auth but it's still 400kb in total. Would love to see the modularised AWS SDK used in this preview version.

**Comment by jadbox** [comment]
I only use auth, but this branch still pulls in Analytics in my build unfortunately.

**Comment by jadbox** [comment]
@powerful23 is there plans to keep the preview tagged release in-sync with the updates happening in stable? I think the current preview is two releases behind the stable version currently.

**Comment by roblangridge** [comment]
I'm on preview 2.0.0-preview.1 and i only use Auth & Analytics, when i do:

import { Analytics } from "aws-amplify";

i still get GraphQL, PubSub and all of aws-sdk

<img width="552" alt="Screenshot 2019-07-11 at 20 34 57" src="https://user-images.githubusercontent.com/15785861/61079853-a755c880-a41b-11e9-93bf-b7ea6a3bb33a.png">

**Comment by powerful23** [comment]
@jadbox @roblangridge are you running the app in the production mode?

**Comment by roblangridge** [comment]
@powerful23 yes, with mode: "deployment" things are looking like this:

<img width="555" alt="Screenshot 2019-07-11 at 21 09 37" src="https://user-images.githubusercontent.com/15785861/61081901-441a6500-a420-11e9-9200-f5db10f42451.png">

**Comment by jadbox** [comment]
@powerful23 not sure what you mean by "production mode", but building with CRA via "npm build" is what I'm using, while setting NODE_ENV to production.

I can still confirm that graphql is being bundled even though I'm not importing it or any libs that use it:
![19-07-12-09-16](https://user-images.githubusercontent.com/595843/61143198-63e47400-a486-11e9-9b1c-ba63a084a6f4.png)

If need be, I can get you access to my project for testing (jdunlap AT outlook.com).

**Comment by powerful23** [comment]
@jadbox I think you need to configure your webpack config to add `mode: "production"` in it. https://webpack.js.org/guides/production/#specify-the-mode

**Comment by aldo-roman** [comment]
@jadbox  or better yet, how are you building? If you are using `create-react-app` then just running `npm run build` activates de production mode. But this may not be the case if you have a manual `webpack.config.js` or some other bundler.

**Comment by davidmold** [comment]
It's really great, but seems poorly documented, that if you just want the Auth module, and you do this:

import Auth from '@aws-amplify/auth'

You save nearly 600K on your build over doing this:

import { Auth } from 'aws-amplify'

**Comment by jadbox** [comment]
@powerful23 @aldo-roman sorry for the delay, I am building in production mode (via create-react-app). All my imports are in the style of `import { Auth } from 'aws-amplify'`.

Using `react-scripts v3.0.1` and `aws-amplify: 2.0.0-preview.1`

**Comment by hgl** [comment]
I'm having difficulty bundling aws-amplify with rollup, encountering the exact same error described in https://github.com/aws/aws-sdk-js/issues/1769#issuecomment-339655699

I think because aws-sdk v2 isn't published as es modules.

Could aws-amplify depend on v3 instead v2 to make it be more friendly to modern bundler like rollup?

**Comment by roborourke** [comment]
I've ended up rolling my own analytics and auth script using the v3 sdk as I wasn't really using any other parts of amplify. Good news is that bundle is a 35kb download. It's part of a WP plugin right now but I'll see if I can extract it out as an npm package.

**Comment by hgl** [comment]
@roborourke how do you roll your own auth with v3 sdk? afaik, the only relevant package is client-cognito-identity-browser, and it only lets you manage the user pool & identity pool but not the actual authorization?

**Comment by roborourke** [comment]
@hgl you call GetId which requires no auth, then GetCredentialsForId with that ID. Then you can use those credentials for calling the PutEvents and UpdateEndpoint pinpoint commands

**Comment by hgl** [comment]
@roborourke Thanks for the heads up. If I'm not wrong, this is the identity pool auth flow, which means you have to log in somewhere to get the id token before you can retrieve the identity id? I guess in my case I still have to log in the user pool and then call GetId?

**Comment by artecoop** [comment]
I strongly agree with those who ask for a better support for bundlers. Behind webpack, I'd love to use Rollup or something else.
For the next iteration, please include also aws-sdk v3!!!!

**Comment by jadbox** [comment]
Btw there's another PR for modularization https://github.com/aws-amplify/amplify-js/pull/3734

**Comment by jonoirwinrsa** [comment]
Changing to the preview branch took my bundle size from 5.24Mb to 2.82Mb! 👏 

Still getting quite a lot of the aws-sdk but far less of it now
![Screenshot 2019-08-03 at 11 22 36](https://user-images.githubusercontent.com/5059404/62410087-2f755b00-b5e1-11e9-8386-b843c1fc8c2e.png)

**Comment by nvrk** [comment]
FYI, `aws-amplify@preview` breaks with SSR on NextJS –  yields `window` undefined.
However, `aws-amplify@unstable` seems to work fine.

**Comment by henrysachs** [comment]
In my Angular Production build when disabling debug logging i get PubSub is not configured but i dont even use PubSub. I added PubSub.configure() but that doesn't seem to help. 

My Setup is Auth, Storage, Graphql

In the Debug logging i can see that Auth doesn't seem to be configured but the login works o.O 

If someone could help me debug this further i would be happy to provide the necessary Information that may be still missing.

**Comment by tafelito** [comment]
preview version seems is fixed at 1.1.31-preview.41 and latest release is at 1.1.40. Does latest version includes changes made for preview? Docs are so confusing

**Comment by pravgupt** [comment]
@tafelito The preview release is not up to date and the docs are also not updated for it's consumption. Latest version does not have changes made in preview.

We are currently working on this RFC, please stay tuned for more details.

**Comment by sammartinez** [comment]
#634

**Comment by sammartinez** [comment]
#1336

**Comment by ajhool** [comment]
In the meantime, a recommended approach for dynamically importing (webpack and/or next.js) individual libraries from the client would be appreciated immensely. Is @tuhinkarmakar  's approach really the best way to do it?

**Comment by depiction** [comment]
Any update on this? When will this be released?

**Comment by rickcha** [comment]
For those who use auth only, I found this way to reduce the bundle size significantly.
```
import Amplify from "@aws-amplify/core";
import Auth from "@aws-amplify/auth";
Amplify.configure({ auth: { ... } })
```

**Comment by sw-yx** [comment]
i think this is critical particularly for the React developer experience, which i imagine is important.

**Comment by bezreyhan** [comment]
Thanks for prioritizing this change. We are currently using `@aws-amplify/auth` and it is adding 311kb minified / 86kb gzipped. This is way bigger than our app can afford. Two questions:
1. Do we have an ETA for when the modularization will be finished?
2. Any idea what the new bundle size will be?

Alternatively, we've been thinking about migrating to using `amazon-cognito-identity-js`, which is 14kb gzipped. If we are just using `@aws-amplify/auth`, what is the biggest down side to migrating to `amazon-cognito-identity-js`?

**Comment by jliebrand** [comment]
Any update on (a) when this will be merged in to the main release, and (b) when GraphQL will be removed from those needing only the API ?

**Comment by sammartinez** [comment]
Hey everyone, wanted to provide an update of the Modularization work, we currently heads down on this in order to deliver this feature in the coming weeks. Will provide an update on this next week for you. Thanks!

**Comment by Amplifiyer** [comment]
@jliebrand and @bezreyhan, we are in our final phase of testing the new modularization release and is available for you to test as well using [preview tag](https://www.npmjs.com/package/aws-amplify/v/1.1.31-preview.102). Can you try testing that release in the meantime and let us know if you face any issues?

**Comment by bezreyhan** [comment]
> @depiction, @jliebrand and @bezreyhan, we are in our final phase of testing the new modularization release and is available for you to test as well using [preview tag](https://www.npmjs.com/package/aws-amplify/v/1.1.31-preview.102). Can you try testing that release in the meantime and let us know if you face any issues?

I'm using `@aws-amplify/auth`. What would be the best way to test out the preview tag?

**Comment by Amplifiyer** [comment]
> I'm using `@aws-amplify/auth`. What would be the best way to test out the preview tag?

You can use the npm preview tag as `yarn add @aws-amplify/auth@preview` All the current published tags for aws-amplify can be seen [here](https://www.npmjs.com/package/aws-amplify/v/1.1.31-preview.102?activeTab=versions)

**Comment by Amplifiyer** [comment]
# Modularization is available in Preview

Hi all,

With us launching Modularization in the coming weeks, we wanted to provide the ability for you to use our current preview tag (https://www.npmjs.com/package/aws-amplify) in order to start using Modularization within your applications. We **highly recommend to not use this preview tag in production** as we are still in the process of fixing small bugs and validating with our own testing internally.

## Breaking Changes:

Below is a breakdown of the breaking changes that are coming with v3 that we wanted to make you aware of:

* AWS.credentials and AWS.config don’t exist anymore anywhere in Amplify JS
    * Both options will not be available to use in version 3. You will not be available to use and set your own credentials. Migration plan on “How to migrate to using Amplify provided credentials” will follow in the coming weeks after GA launch.
* v2 of AWS SDK JS is not vended as part of Amplify JS. If you rely on v2 of AWS SDK JS, you will need to export it separately. Since AWS SDK JS v3 is exported modularly, only AWS Services that Amplify JS currently supports will be available in v3.
* If you are using exported paths within your Amplify JS application, (e.g. import from @aws-amplify/analytics/lib/Analytics this will now break and no longer will be supported. You will need to change to regular import statements.
* If you are using categories as Amplify.<Category>, this will no longer work and we recommend to import the category you are needing to use.
    * The one exception to this is using aws-amplify-angular package where we are still supplying all instantiated categories. This is subject to change in a later release.
* For aws-amplify-react for the Authenticator Component, you will need to import the styles yourself.
    * You will need to import the following: `import '@aws-amplify/ui/dist/style.css';`



## Installing preview-tag:

### For installing Amplify modules:
You are able to install by running `yarn add @aws-amplify/<category>@preview`. E.g. `yarn add @aws-amplify/auth@preview` if you are using the Auth Module. 

### For installing aws-amplify:
You are able to install by running `yarn add aws-amplify@preview`

### For installing UI Components:
You are able to install by running `yarn add aws-amplify-<framework>@preview`. E.g. `yarn add aws-amplify-react@preview` if you are using our React UI Components.

**NOTE**: Please delete your existing *node_modules* before installing as there are some conflicts with the older versions of package.

## Filling Bugs:

If you do find issues, please let us know by cutting us a [bug](https://github.com/aws-amplify/amplify-js/issues/new?assignees=&labels=to-be-reproduced&template=bug_report.md&title=) and we can look into the issue.

Thank You!
Amplify JS team

**Comment by 0xdevalias** [comment]
Not sure if actually a bug, or my own issue. Can you provide a minimal example of how to use `withOAuth` to use the Cognito hosted login?

This is what I have, and but I get a pile of 'cannot import' type errors. Installed both with the `@preview` version, and deleted my `node_modules` first:

```
"aws-amplify": "^1.1.31-preview.110",
"aws-amplify-react": "^2.3.11-preview.123",
```

```javascript
import { Auth } from 'aws-amplify'
import { withOAuth } from 'aws-amplify-react'

Auth.configure({
  // REQUIRED - Amazon Cognito Region
  region: 'us-east-1',

  // OPTIONAL - Amazon Cognito User Pool ID
  userPoolId: 'us-east-1_REDACTED',

  // OPTIONAL - Amazon Cognito Web Client ID (26-char alphanumeric string)
  userPoolWebClientId: 'REDACTED',

  // OPTIONAL - Enforce user authentication prior to accessing AWS resources or not
  mandatorySignIn: false,

  // Cognito Hosted UI configuration
  oauth: {
    domain: 'REDACTED',
    scope: ['email', 'profile', 'openid'],
    redirectSignIn: 'http://localhost:3000/',
    redirectSignOut: 'http://localhost:3000/',
    responseType: 'code', // or 'token', note that REFRESH token will only be generated when the responseType is code
  },
})
```

![image](https://user-images.githubusercontent.com/753891/77148652-107df780-6ae4-11ea-91e9-4c1b36c32d2e.png)
(etc)

**Comment by Amplifiyer** [comment]
Hi @0xdevalias, there shouldn't be any difference between installing both the packages in one command or separately, it seems that aws-amplify-react package was installed the preview version but somehow aws-amplify was reverted or still using the non-preview version. Let us know if you are able to reproduce it again.

**Comment by ImedAdel** [comment]
@Amplifiyer can I suggest one more thing? Moving `aws-sdk` to peer dependencies :)

**Comment by Amplifiyer** [comment]
@ImedAdel Do you mean aws-sdk version 2.x or the new aws-sdk-js-v3? We have completely removed the dependency on aws-sdk version 2.x and only depends on aws-sdk-js-v3. I'm not completely convinced that moving that to peer dependency is better mostly because currently the whole Amplify library extensively depends on it and have strict requirements on which version to use.

Since peer dependencies don't help in reducing the application bundle size but only library's bundle size which is mostly irrelevant (that's my understanding, happy to be convinced otherwise).

**Comment by Amplifiyer** [comment]
Closing this RFC as all the proposed solutions have been launched, see https://aws.amazon.com/blogs/mobile/amplify-framework-announces-new-rearchitected-ui-component-and-modular-javascript-libraries/

Thank you everyone for great feedback and all the contributors for making this happen.
